llvm/lib/Transforms/Utils/CodeExtractor.cpp

```diff
@@ -1756,140 +1439,227 @@ CodeExtractor::extractCodeRegion(const CodeExtractorAnalysisCache &CEAC,
       EntryFreq +=
           BFI->getBlockFreq(Pred) * BPI->getEdgeProbability(Pred, header);
     }
-  }
 
-  // Remove @llvm.assume calls that will be moved to the new function from the
-  // old function's assumption cache.
-  for (BasicBlock *Block : Blocks) {
-    for (Instruction &I : llvm::make_early_inc_range(*Block)) {
-      if (auto *AI = dyn_cast<AssumeInst>(&I)) {
-        if (AC)
-          AC->unregisterAssumption(AI);
-        AI->eraseFromParent();
+    for (BasicBlock *Succ : ExtractedFuncRetVals) {
+      for (BasicBlock *Block : predecessors(Succ)) {
+        if (!Blocks.count(Block))
+          continue;
+
+        // Update the branch weight for this successor.
+        BlockFrequency &BF = ExitWeights[Succ];
+        BF += BFI->getBlockFreq(Block) * BPI->getEdgeProbability(Block, Succ);
       }
     }
   }
 
+  // Determine position for the replacement code. Do so before header is moved
+  // to the new function.
+  BasicBlock *ReplIP = header;
+  while (ReplIP && Blocks.count(ReplIP))
+    ReplIP = ReplIP->getNextNode();
+
+  // Construct new function based on inputs/outputs & add allocas for all defs.
+  std::string SuffixToUse =
+      Suffix.empty()
+          ? (header->getName().empty() ? "extracted" : header->getName().str())
+          : Suffix;
+
+  ValueSet StructValues;
+  StructType *StructTy;
+  Function *newFunction = constructFunctionDeclaration(
+      inputs, outputs, EntryFreq, oldFunction->getName() + "." + SuffixToUse,
+      StructValues, StructTy);
+  newFunction->IsNewDbgInfoFormat = oldFunction->IsNewDbgInfoFormat;
+
+  emitFunctionBody(inputs, outputs, StructValues, newFunction, StructTy, header,
```

Comment:
after this commit, we're seeing msan failures saying there's a use of uninitialized memory on this line

```
    #0 0x7f1a2b16a7a1 in llvm::CodeExtractor::extractCodeRegion(llvm::CodeExtractorAnalysisCache const&, llvm::SetVector<llvm::Value*, llvm::SmallVector<llvm::Value*, 0u>, llvm::DenseSet<llvm::Value*, llvm::DenseMapInfo<llvm::Value*, void>>, 0u>&, llvm::SetVector<llvm::Value*, llvm::SmallVector<llvm::Value*, 0u>, llvm::DenseSet<llvm::Value*, llvm::DenseMapInfo<llvm::Value*, void>>, 0u>&) llvm-project/llvm/lib/Transforms/Utils/CodeExtractor.cpp:1474:3
    #1 0x7f1a2b168abf in llvm::CodeExtractor::extractCodeRegion(llvm::CodeExtractorAnalysisCache const&) llvm-project/llvm/lib/Transforms/Utils/CodeExtractor.cpp:1377:10
    #2 0x7f1a2f777242 in llvm::OpenMPIRBuilder::finalize(llvm::Function*) llvm-project/llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:719:38
    #3 0x7f1a3522aa86 in (anonymous namespace)::OpenMPIRBuilderTest_ParallelNested_Test::TestBody() llvm-project/llvm/unittests/Frontend/OpenMPIRBuilderTest.cpp:921:14
```

Reply:
@Meinersbur

Reply:
I suspect it to be `StructTy` which is not initialized if not using a struct. Would not be used in that case either. Working on a fix...

Reply:
This was fixed already in 9b6b9d39030f92ea20d8f38ef37305e99cfc6d60 which was committed before your comment.

It would be good if you had linked the buildbot.

Reply:
sorry, this was in some internal non-public bots. yes that commit fixes this, thanks for the pointer